### DISCUSSION TITLE
How to get the mesh data of the scene for rendering data?

### DISCUSSION THREAD

**Discussion by WangKai1007** [StereoKit/StereoKit#360] [Q&A] [answered]
I  use StereoKit to develop an openXR app for an projector device. Here is my question.
I  have a  mesh data,like cube or tree model，in scene . And then ,I use the mouse event to rotate or translate the mesh.
There is only part of the mesh showed in the scene.How can I get the data of the  part of the mesh showed in the scene. 
 I want to use the part of the 3D data (Mesh or vertex )for other purpose.  I think the data of part of the mesh is stored in rendering engine.But I don`t know how to get it.

**Comment by maluoi** [comment]
StereoKit does have a `Mesh.GetInds` and `Mesh.GetVerts`! These functions are a little slow, so definitely don't call them often, but will absolutely work for grabbing the geometry data!

**Reply by WangKai1007** [reply]
Thank you for your reply.  
I don`t want to get the whole data  of the mesh.  I just want to get the part of the mesh in the scene.
For example, A tree model in the scene. I see it from the top view， I just can see the leaves, branches, but no trunk. So How can I get the data of the leaves, and branches.this is my purpose.

**Reply by maluoi** [reply]
This sounds pretty complicated, and raises quite a few questions!
- Does it need to be real-time?
- Are the leaves solid meshes, or quads with textures? Basically, is transparent geometry of any kind involved?
- Are you testing occluded visibility of the entire mesh or mesh bounds, or do you want per-triangle, or per-pixel visibility?
- Can you do something else that is simpler, but achieves the same effect?

Perhaps a little explanation of what exactly you're trying to accomplish would be helpful!

**Comment by WangKai1007** [comment, marked as answer]
Maybe I did not express it clearly.  I have a Camera，and  I have the camera all parameter,internal parameters [Internal Matrix], camera external parameters [external Matrix], and ccd size（1280,720）,distortion is not considered.  I want to use this camera to capture the mesh.  The vertex in mesh is pointrgb(Xw,Yw,Zw,r,g,b). so I can the get  picture from the equation.   
zoomfactor*[u,v]=[Internal Matrix]*[external Matrix](Xw,Yw,Zw).
I think this is the simpler way to the capture ,but it is very time-consuming. and This way can not distinguish the data in sight (we can see in the scene) and the data not in sight. That is why I want to find the method from the rendering engine.I think the engine can show the data in sight,maybe  it can store the data in somewhere.
for your quesion;
1 At preset ，real-time is not first concern. Make it work firstly.
2 leaves or object is mesh with color or points with RGB,they all opaque.
3 I want to test occluded visibility the mesh,by per-triangle, or per-pixel visibility. I want to find all occluded visibility vertex in the mesh and  then use the equation to calculate the image.
4 Actually,I can set a radial line  from camera to target .And then use the radial line to intersect with the mesh, to get the visible point.
this is very bad method.

**Reply by WangKai1007** [reply]
Since ，I still can not  get the part mesh data of total. I use  Renderer.Screenshot(). to instead. 
I still need  other way to get  3D data .

**Reply by maluoi** [reply]
This sounds like a really tricky task all up! It sounds like the sort of thing that compute shaders could be good at, but I'm not even sure how you'd go about doing that. SK doesn't directly support compute shaders right now either.

Another approach could be to draw using a stencil buffer to store an id. You might be a little limited in the number of ids, but SK doesn't support stencils yet either! You could apply essentially the same strategy, drawing to different color buffer types, write an id instead of a color, and then reading it back to CPU to see what's there.

**Reply by WangKai1007** [reply]
The  function ,"Render.Screenshot()",is just way around that. As you said，It is something about computer shaders. I just can not find a fast  and  efficient way to  get it .

**Reply by paulmelis** [reply]
As a curious bystander for this issue, how much are you tied to the XR device in your use case? Would you able to use a second renderer to get the data you want? E.g. send the camera parameters to a PC over wifi, and have it do the intersection processing on the same scene you need instead of doing it on the device itself?
